server/encryptionkm/key_manager.go

```diff
@@ -14,41 +14,469 @@
+	}
+	// Avoid write conflict with PD peer by checking if we are leader.
+	resp, err := leadership.LeaderTxn().
+		Then(clientv3.OpPut(EncryptionKeysPath, string(value))).
+		Commit()
+	if err != nil {
+		log.Warn("fail to save encryption keys.", zap.Error(err))
+		return errs.ErrEtcdTxn.Wrap(err).GenWithStack("fail to save encryption keys")
+	}
+	if !resp.Succeeded {
+		log.Warn("fail to save encryption keys. leader expired.")
+		return errs.ErrEncryptionSaveDataKeys.GenWithStack("leader expired")
+	}
+	// Leave for the watcher to load the updated keys.
+	log.Info("saved encryption keys")
+	return nil
+}
+
+func loadKeysFromKV(kv *mvccpb.KeyValue) (*encryptionpb.KeyDictionary, error) {
+	content := &encryptionpb.EncryptedContent{}
+	err := content.Unmarshal(kv.Value)
+	if err != nil {
+		return nil, errs.ErrProtoUnmarshal.Wrap(err).GenWithStack(
+			"fail to unmarshal encrypted encryption keys")
+	}
+	masterKeyConfig := content.MasterKey
+	if masterKeyConfig == nil {
+		return nil, errs.ErrEncryptionLoadKeys.GenWithStack(
+			"no master key config found with encryption keys")
+	}
+	masterKey, err := encryption.NewMasterKey(masterKeyConfig)
+	if err != nil {
+		return nil, err
+	}
+	plaintextContent, err := masterKey.Decrypt(content.Content, content.Iv)
+	if err != nil {
+		return nil, err
+	}
+	keys := &encryptionpb.KeyDictionary{}
+	err = keys.Unmarshal(plaintextContent)
+	if err != nil {
+		return nil, errs.ErrProtoUnmarshal.Wrap(err).GenWithStack(
+			"fail to unmarshal encryption keys")
+	}
+	return keys, nil
+}
 
 // NewKeyManager creates a new key manager.
-func NewKeyManager(kv kv.Base, config *lib.Config) (*KeyManager, error) {
-	// TODO: Implement
-	return &KeyManager{}, nil
+func NewKeyManager(
+	ctx context.Context,
+	etcdClient *clientv3.Client,
+	config *encryption.Config,
+) (*KeyManager, error) {
+	return newKeyManagerImpl(ctx, etcdClient, config, defaultKeyManagerHelper())
+}
+
+// newKeyManager creates a new key manager, and allow tests to set a mocked keyManagerHelper.
+func newKeyManagerImpl(
+	ctx context.Context,
+	etcdClient *clientv3.Client,
+	config *encryption.Config,
+	helper keyManagerHelper,
+) (*KeyManager, error) {
+	method, err := config.GetMethod()
+	if err != nil {
+		return nil, err
+	}
+	masterKeyMeta, err := config.GetMasterKeyMeta()
+	if err != nil {
+		return nil, err
+	}
+	m := &KeyManager{
+		etcdClient:            etcdClient,
+		method:                method,
+		dataKeyRotationPeriod: config.DataKeyRotationPeriod.Duration,
+		masterKeyMeta:         masterKeyMeta,
+		helper:                helper,
+	}
+	// Load encryption keys from storage.
+	_, err = m.loadKeys()
+	if err != nil {
+		return nil, err
+	}
+	// Start periodic check for keys change and rotation key if needed.
+	go m.startBackgroundLoop(ctx, m.mu.keysRevision)
+	return m, nil
+}
+
+func (m *KeyManager) startBackgroundLoop(ctx context.Context, revision int64) {
```

Comment:
Can we name it as `parCtx` instead of `ctx`.

Reply:
We need to use `serverLoopCtx` and `serverLoopWg` here. Please follow the comment in `server.go`.

Reply:
@Yisaer what's `par` stands for?